### ISSUE TITLE
[Feature Request] Native QOI Image Format Support

### ISSUE THREAD

**Issue by Fios-Dev** [libgdx/libgdx#6844]
Please ensure you have given all the following requested information in your report.

#### Issue details
QOI is a new lossless image format with comparable compression ratios to PNG but with significantly faster processing speed due to the simplicity of its compression and decompression algorithms. The actual algorithm is fairly simple to implement as well. I'd offer to do it myself, but I wouldn't know where to start in the codebase.

https://qoiformat.org/

#### Reproduction steps/code
N/A

#### Version of libGDX and/or relevant dependencies
All

#### Stacktrace
N/A

#### Please select the affected platforms
- [ x] Android
- [ x] iOS
- [ x] HTML/GWT
- [ x] Windows
- [ x] Linux
- [ x] macOS

**Comment by Frosty-J** [comment]
I'm not particularly well-versed in this, but https://github.com/libgdx/libgdx/blob/master/gdx/jni/gdx2d/stb_image.h seems to be where image decoding takes place.

Personally, I'd rather see support for more mainstream formats, such as WebP and AVIF, which both achieve better compression ratios than PNG and debatably JPEG (I find WebP smoothens photos too much and doesn't really win over MozJPEG, but AVIF is probably objectively better). But they are complex formats that can't make the same 300-line claim QOI does!

I took a look at the sample images, and PNG usually wins in terms of file size. So this mostly boils down to whether anyone's game is spending a long time decoding images. I haven't had any trouble with that, even without exploring multithreading, but my games are small in scope, so we'd really have to hear from someone else.

I assume this would be slower than ZKTX and whatever other texture formats we have, but also smaller in size to get a nice middle ground between that and PNG?  
https://libgdx.com/wiki/graphics/2d/texture-compression

**Comment by tommyettinger** [comment]
The actual algorithm may be simple to implement, but unless it can be done with some degree of cross-platform support, this becomes less likely. We just dealt with a significant hassle from some file format handling code that was missing bounds checks, and had existed for a while -- obviously that's bad and we want to avoid increasing our vulnerable surface for that type of thing.  We had tried updating stb_image as well, and it's unclear whether the current version works like it should. It looks like there isn't built-in support for QOI in PhotoShop, GIMP, or other common image editors; KTX isn't a widely-supported format either, I suppose, but this doesn't help the usability of QOI. As Frosty-J pointed out, decoding time for images hasn't been a bottleneck for any games I know of. Image file size is usually not huge, either -- music can often use much more space, depending on format and quality. Just bundling a JRE normally adds significantly more size than several 4096x4096 PNG atlases.

I think we should leave this open in case QOI becomes more widely-supported, but I suggest leaving the request to be potentially implemented some time in the future.

**Comment by code-disaster** [comment]
I don't agree on the performance topic. PNG encoding/decoding speed has been *the* bottleneck for building *and* loading texture atlasses with Pathway. I switched to a custom binary format which uses LZ4 compression, trading storage space (about ~ 2x the size of PNG) for better performance and  much less memory overhead.

That said, QOI won't magically solve everything. For example, brief tests indicate that it's actually not very well suited for pixel art style assets. You are also right that there's no widespread content creation tools support yet, though I wouldn't mind that part - I'd primarily use it as a intermediate/target format anyway.

Also, I wouldn't bother doing an own implementation. There are plenty of ports already, including a Java one. I'd probably just create a C wrapper to the reference implementation.

**Comment by tommyettinger** [comment]
I haven't ever built atlases at runtime, so that explains why I've never noticed the encoding/decoding time. LZ4 seems like a good option for very fast encode/decode times. It would be interesting to compare QOI against your custom format -- since QOI emphasizes a simple implementation, it could easily be worse than your LZ4-based solution.

I agree that using an existing implementation would be best, but I have no idea how we would cram another format into libGDX, since stb_image seems to be doing most of the work. I also suspect we don't want to cram another format into libGDX while the existing stb_image can't be updated, last I checked, for stability reasons. Considering that stb_image is tested very thoroughly, that QOI probably has not been tested to the same extent, and that we still had issues with stb_image is not a vote of confidence.

**Comment by Frosty-J** [comment]
https://github.com/richgel999/fpng might be worth checking out. Claims to be 12-19× faster at encoding than stb_image and 2.5-3× faster at decoding. Obviously still a smaller claim than QOI's and probably still worse than LZ4.

There's also https://github.com/libjxl/libjxl/tree/main/experimental/fast_lossless for JPEG XL encoding, but we don't support JPEG XL and the browsers still have it locked behind flags, so not so useful.

**Comment by code-disaster** [comment]
FPNG looks promising, but has one major restriction: their decoder only works on png files encoded by FPNG.
